Jenkinsfile

```diff
@@ -111,91 +113,11 @@ def functionalDynamicParallelSteps(image){
-        def currentStep = i;
-        ciGroupsMap["${currentCiGroup}"] = {
-            stage("${currentCiGroup}") {
-                withEnv([
-                    "TEST_BROWSER_HEADLESS=1",
-                    "CI=1",
-                    "CI_GROUP=${currentCiGroup}",
-                    "GCS_UPLOAD_PREFIX=fake",
-                    "TEST_KIBANA_HOST=localhost",
-                    "TEST_KIBANA_PORT=6610",
-                    "TEST_ES_TRANSPORT_PORT=9403",
-                    "TEST_ES_PORT=9400",
-                    "CI_PARALLEL_PROCESS_NUMBER=${currentStep}",
-                    "JOB=ci${currentStep}",
-                    "CACHE_DIR=${currentCiGroup}"
-                ]) {
-                    image.inside {
-                        sh "node scripts/functional_tests.js --config test/functional/config.js --include ${currentCiGroup}"
-                    }
-                }
-            }
-        }
-    }
-    parallel ciGroupsMap
-    ciGroupsMap = [:]
-    for (int i = 7; i <= 9; i++) {
-        def currentCiGroup = "ciGroup${i}";
-        def currentStep = i;
-        ciGroupsMap["${currentCiGroup}"] = {
-            stage("${currentCiGroup}") {
-                withEnv([
-                    "TEST_BROWSER_HEADLESS=1",
-                    "CI=1",
-                    "CI_GROUP=${currentCiGroup}",
-                    "GCS_UPLOAD_PREFIX=fake",
-                    "TEST_KIBANA_HOST=localhost",
-                    "TEST_KIBANA_PORT=6610",
-                    "TEST_ES_TRANSPORT_PORT=9403",
-                    "TEST_ES_PORT=9400",
-                    "CI_PARALLEL_PROCESS_NUMBER=${currentStep}",
-                    "JOB=ci${currentStep}",
-                    "CACHE_DIR=${currentCiGroup}"
-                ]) {
-                    image.inside {
-                        sh "node scripts/functional_tests.js --config test/functional/config.js --include ${currentCiGroup}"
-                    }
-                }
-            }
-        }
-    }
-    parallel ciGroupsMap
-     ciGroupsMap = [:]
-    for (int i = 10; i <= 12; i++) {
-        def currentCiGroup = "ciGroup${i}";
-        def currentStep = i;
-        ciGroupsMap["${currentCiGroup}"] = {
-            stage("${currentCiGroup}") {
-                withEnv([
-                    "TEST_BROWSER_HEADLESS=1",
-                    "CI=1",
-                    "CI_GROUP=${currentCiGroup}",
-                    "GCS_UPLOAD_PREFIX=fake",
-                    "TEST_KIBANA_HOST=localhost",
-                    "TEST_KIBANA_PORT=6610",
-                    "TEST_ES_TRANSPORT_PORT=9403",
-                    "TEST_ES_PORT=9400",
-                    "CI_PARALLEL_PROCESS_NUMBER=${currentStep}",
-                    "JOB=ci${currentStep}",
-                    "CACHE_DIR=${currentCiGroup}"
-                ]) {
-                    image.inside {
+                    image.inside("-v \'${env.WORKSPACE}_${currentCiGroup}:${env.WORKSPACE}/optimize\'") {
```

Comment:
Let's see if you can use the last approach. By creating multiple empty directories in the workspace each name after corresponding ciGroups. Then mount each directory to corresponding CI group while running functional test in Jenkins.